Allow reloading tools without restarting galaxy, simplify tool panel handling for tool shed installed tools

mvdbeek opened the pull request:
This PR adds the ability to install new tools and data managers into galaxy without needing to restart galaxy, when galaxy is run in production mode (meaning multiple handler threads).
The key is to have the toolconf watcher send a message to the internal message queue, which will cause all handlers to reload the toolbox.
The shed_tool_conf.xml is now being treated exactly like the tool_conf.xml, meaning as soon as any changes are made to that file (for example when installing new tools from the toolshed), galaxy will reload the toolbox, and all handers will have access to the new toolbox.

This PR also changes the behaviour of deactivating repositories. Previously, the deactivated tool would remain in the shed_tool_conf.xml, so that re-installing the repository would place the tool (in addition to the section that has been chosen when re-installing the tools) into the section it was previously installed. We now simply remove the tool (The previous behaviour has never been documented anywhere AFAIK).

mvdbeek committed:
Try reloading tool via control_task

dannon commented:
@mvdbeek +1 to what's here, but also, how about we just automatically kick this message on new tool installation?

(edit:  And, no, you don't need to do any extra queue configuration -- it should work out of the box, though you can certainly configure a more robust queue if you want)

mvdbeek commented:
> how about we just automatically kick this message on new tool installation?

absolutely, but I haven't figured out yet where, and how this currently handled.

dannon commented:
@mvdbeek Believe it or not, this was one of the first things I tried to do when I first wrote all of this queue stuff.  At the time, it was a bit of a mess and the tool install mechanisms were going to be refactored to where there was a single point we could hook to for a reload, but I don't know the state of that -- haven't looked at it in a long time.

jmchilton commented:
For non-shed tool conf each Galaxy process will just monitor and reload on changes. If the shed process with respect to these configs can be simplified to just adding an entry to shed_tools_conf.xml (no more updating internal data structures related to the panel for instance) then we can just have Galaxy monitor shed tool confs also and reload on change. There would be no need to for this coordination. If you are going to dig into this - that would be my preferred approach to this problem.

jmchilton commented:
^ That was for the long term - I just want shed and non-shed tools to act the same. This PR is definitely good and is desirable regardless of that - thanks for the awesome work @mvdbeek!

mvdbeek commented:
> For non-shed tool conf each Galaxy process will just monitor and reload on changes. If the shed process with respect to these configs can be simplified to just adding an entry to shed_tools_conf.xml (no more updating internal data structures related to the panel for instance) then we can just have Galaxy monitor shed tool confs also and reload on change. There would be no need to for this coordination. If you are going to dig into this - that would be my preferred approach to this problem.

That does sounds like a good idea. I'll see if I can come up with something in that direction.

mvdbeek committed:
Explicitly monitor shed_tool_conf for changes

mvdbeek committed:
Reference app instead of self when reloading toolbox

mvdbeek committed:
Do not load individual new toolpanel elements (We rely on the tool watcher to reload the toolbox)

mvdbeek committed:
Remove in-memory removal of tools (rely on toolbox reload)

mvdbeek committed:
Remove one more unused panel manipulation function

mvdbeek committed:
Remove unused variable

mvdbeek committed:
Start conf_watcher thread in unit tests

mvdbeek committed:
Wait before triggering a callback in test_watcher.py

mvdbeek committed:
Remove update_shed_conf test

We do not update the toolbox in memory, so this test is obsolete

mvdbeek committed:
Remove panel_manager tests for deactivated tools

Deactivate tests that rely on deactivating, but not removing entries
from tool_conf. For tests that test changes in tool_panel, load a new
toolbox after making the changes.

mvdbeek committed:
Add sleep to allow toolbox to finish loading

mvdbeek committed:
Remove test for deactivated tools and sections

These are tests that verify that deactivated tools were present in the
correct tool panel section.

mvdbeek commented on test/unit/tools/test_watcher.py:
```diff
@@ -35,11 +36,12 @@ def test_tool_conf_watcher():
 
     callback = CallbackRecorder()
     conf_watcher = watcher.get_tool_conf_watcher(callback.call)
+    conf_watcher.start()  # Need to manually start conf_watcher thread
```
This and the sleep should fix https://github.com/galaxyproject/galaxy/issues/2831

nsoranzo replied:
Thanks @mvdbeek! I'll test when I'm back from holiday.

natefoo replied:
@mvdbeek I just found this, I fixed it slightly differently in #2856

mvdbeek committed:
Wait 2 seconds after touch shed_tool_conf.xml

The file watcher monitors for changes every second, so 2 seconds should
be enough to let tests that check the tool panel after installing new
tools pass.

mvdbeek committed:
Also reload data managers upon touching data_manager.conf

mvdbeek commented:
Okay @dannon @jmchilton @martenson, I think this is ready for a first round of review.

I updated the PR description at the top. One slightly annoying thing is that something in the data manager installation process touches the shed_tool_conf.xml, so that also triggers the complete toolbox reload. Oh well, no big problem, I guess we could hash the file and compare hashes when the update time differs.

The other things is that the old toolbox doesn't `shutdown()` properly.
Removing register_postfork_function from https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/tools/toolbox/watcher.py#L156 solves that issue, but I guess that would cause other problems.

mvdbeek commented:
let me know if a docker image would be handy for testing!

mvdbeek committed:
Shut down old data manager watchers

mvdbeek committed:
Implement hashing for tool conf watcher

If modification time of watched files changes, only trigger event if
hashes differ as well

mvdbeek committed:
Log who's triggering a reload

mvdbeek commented:
A few more problems that need solving:
- ~~if the section label is not also a valid section_key, tools get installed outside of the toolpanel
  (and not at all visible until restart)~~
  - ~~Installing a regular tool after a data manager tool will make the data manager tool appear only sporadically in the web interface (until restart). I suspect there may be a problem if uwsgi is using threads (I've got main.1 and main.2).~~

mvdbeek committed:
Add back integrated_tool_panel saving

mvdbeek committed:
Directly reload toolbox or data manager when watcher triggers reload

mvdbeek committed:
Use the tool_data_tables.tool_data_path, to avoid persisting the temporary in-memory path to the data table

mvdbeek committed:
Fix reload thread for uwsgi workers.

and fix typos

mvdbeek committed:
Do not call start() method of watchter

as this happens when it starts watching a file.

mvdbeek committed:
One last attempt to revive launched thread

mvdbeek committed:
Transplant tool conf watcher thread

mvdbeek committed:
Also reload conf_watchers for data managers

mvdbeek committed:
Reload data managers after reloading the toolbox

mvdbeek committed:
Trigger event with lock

mvdbeek committed:
Reload data table managers and data tables

... before reloading data managers. Data managers need to have access to
the data tables, so that they can process the results.
Do not explicitly start watcher thread for shed_data_manager_conf
changes.

jmchilton commented:
Watching the toolbox would't affect data managers I would assume - so it is unsurprising that only certain web threads see them. It doesn't sound like a regression though right?

jmchilton commented on lib/galaxy/tools/toolbox/watcher.py:
```diff
@@ -90,15 +92,24 @@ def check(self):
                 if new_mod_time != mod_time:
-                    self.paths[path] = new_mod_time
-                    do_reload = True
+                    if hashes[path] != md5_hash_file(path):
```
Awesome sauce!

jmchilton commented:
Awesome! +1 from me - let me know if it is ready to merge.

mvdbeek commented:
> Awesome! +1 from me - let me know if it is ready to merge.

Almost! I'm still fighting with data managers that bring their own table.loc.  I could short-circuit the logic and just write to the tool-data/table.loc file, instead of tool-data/toolshed.../..../..../ .
I definitely solved the uwsgi problem though. 
The trick is to ["transplant" the watcher thread to the new toolbox](https://github.com/galaxyproject/galaxy/pull/2840/commits/199437dbae8f3ce81daddf0607ff74f66b3513ec) , instead of creating a new thread (which never runs, I guess because we run uwsgi in preforking mode).  Does that make sense?

Now all web threads always see new tools / data managers.

mvdbeek committed:
Remove debugging statements and mention uwsgi/threads in line

mvdbeek commented:
Yay, data managers problem solved! I'll have a look at the performance when a whole bunch of tools is installed. But other then that I consider it ready.

mvdbeek committed:
Wait for shed_data_manager_conf watcher to pick up changes

mvdbeek commented:
I have done a large-scale test. I fetched a list of all installed repos from usegalaxy and tried to install this (without the tooldeps!) on a VM with 4 cores, with uwsgi running with 4 processes and 4 threads ... it seemed to go not too bad, but I had 19 exceptions like this

```
Exception in thread Thread-136:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "lib/galaxy/tools/toolbox/watcher.py", line 147, in on_any_event
    self._handle(event)
  File "lib/galaxy/tools/toolbox/watcher.py", line 150, in _handle
    self.reload_callback()
  File "lib/galaxy/tools/toolbox/base.py", line 82, in <lambda>
    self._tool_conf_watcher = get_tool_conf_watcher( lambda: reload_toolbox(app))
  File "lib/galaxy/queue_worker.py", line 63, in reload_toolbox
    app.reindex_tool_search()
  File "lib/galaxy/config.py", line 845, in reindex_tool_search
    self.toolbox_search = galaxy.tools.search.ToolBoxSearch( self.toolbox, index_help )
  File "lib/galaxy/tools/search/__init__.py", line 41, in __init__
    self.build_index( index_help )
  File "lib/galaxy/tools/search/__init__.py", line 51, in build_index
    for id, tool in self.toolbox.tools():
RuntimeError: dictionary changed size during iteration
```

And also a whole bunch of "Missing required parameter 'shed_tool_conf'."
The total tally of that experiment is 263 tools installed properly, 63 in error.
Not sure this is actually a regression, I'll try this again without my changes.
And it seemed that after this was done the toolbox kept reloading and reloading ...
a restart fixed that, but that I could have done without this PR. 

I guess I could require a lock for every in-memory access to the toolbox and try to rip out a few more things that manipulate the toolbox.

mvdbeek committed:
Try locking load_item

mvdbeek committed:
Keep a count of reloads

mvdbeek committed:
Always wait for reload action to be finished

mvdbeek commented:
OK, with the additional locking and waiting for the toolbox to return I get 1 out of 299 repos in error,
and that one error is actually an nginx timeout. So, let's get this in!
I'm very curious if this hammer approach of reloading the whole toolbox will work for usegalaxy.
A potential enhancement, if reloading the handlers in parallel is too slow, would be to send the reload jobs in batches. But let's not implement intricate solutions to problems that might not even exist ...

martenson commented:
Reviewing and testing now.

mvdbeek committed:
Mock the Lock; fix linting problems

mvdbeek committed:
One more linting fix

mvdbeek committed:
Fix indentation

jmchilton committed:
Add server_name to unit test app's config.

Required for messaging stuff.

jmchilton committed:
Fix test case for tool manager.

jmchilton committed:
Lint for last commit.

jmchilton committed:
Merge remote-tracking branch 'jmchilton/dev' into reload_toolbox_across_handlers

jmchilton committed:
Another linting fix.

jmchilton commented:
@galaxybot test this

jmchilton commented:
These TS test failures due to timeouts are probably legitimate huh - we should figure out if something is hanging.

jmchilton committed:
Fix the infinite loop.

Thanks @mvdbeek!

bgruening commented:
Amazing work guys! I'm waiting for this far to long :)   :tada:

jmchilton commented:
@mvdbeek your contributions to this project continue to astound me, you continue to improve the tool shed experience for users, developers, and deployers by leaps and bounds and we all thank you for it!

martenson commented:
+1 to what John wrote @mvdbeek ! Your contributions are deeply appreciated.

mvdbeek commented:
Thanks a lot @jmchilton @martenson for the kind words.
